\chapter{Flow Cytometry Analysis}
\label{sec:fcs}

Here we talk about flow cytometry data, how we analyze and visualize it; and how we use that analysis alongside with some machine learning tools to classify samples into cancer subtypes.

\section{Flow Cytometry}
Flow cytometry is a technology that allows measurement of biomarkers inside and outside cells on a single cell basis~\cite{flow-cytometry}. The technology can also sort and separate certain cells according to a given criterion~\cite{flow-cytometry-sorting}.

Cell preparation in flow cytometry involves suspension of the cells in a liquid containing biomarker reagents. Reagents are marked antibodies that can be detected by the laser beams in the flow cytometer machine~\cite{flow-cytometry-cell-preparation}. The antibodies are usually marked with a fluorescent label. Each fluorescent marker has a corresponding peak excitation and emission wavelength which can be detected using lasers or lamps available on the flow cytometer machine. The combination of markers has to be chosen such that their corresponding wavelengths have minimal overlap; otherwise they cannot be distinguished from one another due to interference between them.

In a flow cytometer cells flow in a liquid stream one by one, where a lamps or laser beams in conjunction with sensors measure the intensity of reflected light from the cells. These measurements can be in linear or logarithmic space~\cite{practical-flow-cytometry-book}. The measured values depend on the light intensity projected onto cells which can be tuned by changing the voltage of the lasers or lamps. Different wavelengths correspond to different markers, but they might overlap. When the tail of the emission spectrum of a marker overlaps with the main part of the emission spectrum of another marker, it is called spillover as shown in Fig.~\ref{fig:flow-cytometry-spillover}~\cite{flow-cytometry-compensation}.

Compensating for spillover requires a spillover matrix ($SM$). $SP_{i,j}$ shows the percentage that marker $i$ spills over marker $j$. The compensation matrix ($CM$) is then the calculated as the inverse of the spill over matrix. Let $S$ be the true signal value, and $O$ be the observed value. Then we have\footnote{\url{http://bioinformin.net/cytometry/compensation.php}}:

\begin{align}
  &CM = SM^{-1} \nonumber \\
  &S = O \times CM
  \label{fml:fcs-compensation}
\end{align}

\begin{figure}[!ht]
  \centering
  \includegraphics[width=.8\textwidth]{figs/FITC_spillover_onLSRII_s}
  \caption[Flow Cytometry Spillover Effect]{Fluorescence of a green flurochrome (e.g. FITC) is primarily detected by detector A. However as the emission spectrum is relatively broad some of the fluorescence is detected also by detectors B and C. This is called fluorescence spillover and needs to be corrected for otherwise it could compromise detection of other fluorochromes by their appropriate detectors\footnotemark[1].}
  \label{fig:flow-cytometry-spillover}
\end{figure}

\section{Data Preprocessing and Challenges}
Transformation and spillover compensation are the two main phases of raw flow cytometry data preprocessing.

\emph{Transformation}: The measured fluorescent intensities almost exponentially correspond to the number of existing fluorescent markers on or inside the cell. Therefor a proper transformation of the raw data is essential in order to have the data in a linear space. Logarithmic, log-linear hybrid transformation Logicle~\cite{fcs-logicle}, and hyperbolic arcsine~\cite{fcs-arcsineh} are some commonly used transformations. Some studies have compared different transformation techniques and reported their advantages and disadvantages~\cite{fcs-transformation-survey1, fcs-transformation-survey2}.

\emph{Spillover Compensation}: Compensation is done as shown in Formula~\ref{fml:fcs-compensation} and it relies on a given compensation or spillover matrix.

In practice data are produced through time and also maybe in different labs. This means reagent batches are different, and also flow cytometry machines are not necessarily calibrated alike, which also affects compensation matrices. Therefore normalization is a crucial step to make samples comparable~\cite{fcs-normalization}.

\section{High Dimensional Analysis and Visualization}
Manual analysis of flow cytometry data involves \emph{gating}. Researchers use density or scatter plots of one or two selected dimensions  of flow cytometry data in order to visualize and also select some areas on those plots to further investigate cells within the selected area. Visualization and further gating of those selected cells is commonly a next step to the analysis.

Manual gating of cells across several samples is a labor intensive and time consuming process. Not being able to analyze the data in its original higher dimensional space is another disadvantage of manual flow cytometry data analysis.

The rest of this chapter first explains how we extract features from flow cytometry data using flowType~\cite{Aghaeepour2012}. Then we show how RchyOtimyx uses the outcome of flowType to summarize and visualize gating strategies~\cite{Aghaeepour2012a}. Then an improvement to both functionality and performance of both packages, as well as a pipeline using both methods are presented.

\subsection{Cell Population Identification: flowType}
Assume there is a threshold corresponding to each marker/dimension for a given flow cytometry data. For a given marker $\mathbf{M}$ and the threshold $t$, cells are divided into two groups regarding the threshold as shown in Formula ~\ref{frm:flowtype-pops}.

\begin{align}
  \mathbf{M-} &:= \{c_i | \mathbf{M}(c_i) < t \} \nonumber \\  
  \mathbf{M+} &:= \{c_i | \mathbf{M}(c_i) \ge t \}
  \label{frm:flowtype-pops}
\end{align}

$\mathbf{M}(c_i)$ is the observed value of marker $\mathbf{M}$ for cell $c_i$. Identifying a cell population, some markers might be irrelevant. If a marker does not play a role defining a population, we call it neutral. Therefore regarding each marker, there are $3$ possible populations, \emph{i.e.} positive, negative, neutral. Figure~\ref{fig:flowtype-pops} shows all possible populations considering only two markers\footnote{Credit: Figure~1-A of \cite{Aghaeepour2012}}. Note that a cell population identified by more than one marker is the intersection of the two corresponding sets of cells.

\begin{figure}[!ht]
  \centering
  \includegraphics[width=.6\textwidth]{figs/flowtype/pops}
  \caption{Population identification}
  \label{fig:flowtype-pops}
\end{figure}

The flowType method iterates through all possible combinations of cell populations and reports cell count and/or Mean Fluorescence Intensity (MFI) for each cell population. Given $m$ markers, there are $3^m$ possible combinations, and flowType reports all of them. Some possible further analysis using these reported cell counts and/or MFIs are explored in the original publication~\cite{Aghaeepour2012}.

\subsection{Hierarchical Analysis of Cell Populations: RchyOptimyx}
Recent advances in FCM instrumentation and reagents  have enabled high-dimensional analyses to identify large numbers of  cell populations with potentially significant correlations to an external outcome.
However,  studies often fail to characterize the complex relationships between the markers involved in the  identification of these cell populations.
Revealing this information can provide additional insight into the biological characteristics of the populations identified.
The choice of markers for new panels has been a source of ongoing debate, including efforts such as the Human ImmunoPhenotyping Consortium (HIPC), the Federation of Clinical Immunology Societies Federation of Clinical Immunology Societies (FOCiS) sponsored Flow Immunophenotyping Technical Meetings (FITMaN), and the Optimized Multicolor Immunophenotyping Panels (OMIPs) articles \cite{Maecker2012Standardizing,roederer2010omips,mahnke2010omip,chattopadhyay2010omip,wei2011omip,biancotto2011omip,foulds2012omip,murdoch2012omip,eller2012omip,zuleger2012omip,lamoreaux2012omip,preijers2012omip}.
Understanding the relationships between the markers involved in identification of the target cell population and the characteristics of that cell population (\emph{e.g.,} its correlation with a clinical outcome) is fundamental to the design of effective marker panels. 
%Given this possibility, it is important to consider how many parameters must be measured simultaneously to address effectively a particular hypothesis. To arrive at such conclusions, however, high-dimensional, hypothesis-generating experiments, and the tools to analyze and distill them, will be necessary.
For example, one could use a high-dimensional flow or mass cytometry assay to measure a large list of candidate markers.
However, this can result in parsing the cells into (\emph{e.g.,} clinically) redundant subsets \cite{bendall2012deep}.
Excluding these redundancies (\emph{e.g.}, markers less important for prediction of a clinical outcome) will result in a panel of the most clinically relevant markers.


High dimensional FCM data is usually analyzed using a  laborious sequential manual analysis procedure in which a series of thresholds or 2-dimensional polygons (or gates) are applied to histograms or scatter plots of markers (\emph{e.g.,} \cite{perfetto2004seventeen,gattinoni2011human}).
However, manual gates provide little insight into the relative importance of each gate to the final results.
% For example, when the results need to be reproduced using fewer markers (\emph{e.g.,} for sorting, in clinical settings, or for visualization) it is not clear which gates have higher priority.
For example, consider a six color assay with markers named $1$ to $6$.
%The importance of each marker should  be measured to remove any that are not contributing significantly to the correlation of interest (\emph{i.e.}, identify the parent population with six markers that is most similar to the original phenotype). %Bring up importance here not way down as you want to maintain the thread of what the reader is supposed to be considering
%changes, as the least important marker could still be a significant one; strictly you don't *need* (must) to look at each but it is conceivable you only look at one and remove it if its not significant independent of looking at all the rest Nima: yeah, that's how we originally wrote it but neither Mario nor Holger didn't understand the basic problem. I believe it will be easier for them if we start by stating the obvious and then defining the problem.
If the expression of each marker is considered to be on, off, or does not matter (\emph{e.g.,} markers named $1$, $2$, and $3$ in phenotype $1^+2^-$, respectively), a total of $3^6=729$ cell populations can be distinguished based on these markers.
A given immunophenotype involving all six of these markers (\emph{e.g.,} $1^+2^-3^+4^-5^+6^-$) can have $2^6=64$ parent populations (\emph{e.g.,} $1^+$, $1^+2^-$).
Quantifying the relationship between the cell population of interest and these parent populations is fundamental to our understanding of the importance of the markers for different gating strategies.
The order in which the gates are applied to the data is not important, as long as all of the gates are used (\emph{i.e.,} sequential gating is commutative).
However, to decrease the size of the marker panel, the relative importance of the gates should be determined.
For example, the measurement of the phenotype mentioned above using only five colors requires the determination of the importance of each marker to identify and remove the least important one (\emph{i.e.,} the identification of the parent population with five markers that is most similar to the original phenotype).
This is further complicated by the fact that some cell populations can be identified using more than one combination  of markers and gating strategy; therefore, each marker can be used in different positions in the gating hierarchy and can have different priorities, depending on the choice of the gating strategy.
For example, the $3^+$ gate is involved in both $1^+2^-3^+$ and $3^+4^-5^+$, both parents of the $1^+2^-3^+4^-5^+6^-$ phenotype described above.
However, depending on the amount of redundancy between marker $3$ and others, this marker can have different levels of importance for these two parent populations.

Another use--case for measuring the importance of the markers is  the investigation of a large number of closely related phenotypes (\emph{e.g.}, those identified by bioinformatics pipelines) by identifying their common parent populations.
Several computational tools have been developed for automated identification of cell populations (\emph{e.g.,} \cite{lo2008automated, finak2009merging, pyne2009automated, chan2008statistical, naumann2010curvhdr,  zare2010data, qian2010elucidation, sugar2010misty,  Aghaeepour2010Rapid, bendall2011single, qiu2011extracting}) and 
recent studies have used these tools to identify novel cell populations that correlate with clinical outcomes (\emph{e.g.,} \cite{Aghaeepour2012,zare2012automated,costa2010automated,roederer2011spice,bashashati2012b}).
In addition, the results of the FlowCAP-II project\footnote{http://flowcap.flowsite.org/summit2011.html}  have shown that several algorithms can accurately and reproducibly identify  cell populations correlated with external outcomes. %exceed hasn't been proven as human performance wasn't included in FlowCAp2
However, these algorithms provide limited information regarding the importance of the markers involved in defining the cell populations \cite{Aghaeepour2012,chan2010optimization}. 
This situation is even more complicated than sequential manual gating, since most of these bioinformatics pipelines work based on multivariate classifiers, and as a result, more than one cell population can be responsible for the final predictions.
Therefore, markers can have different relative importance in defining the multiple cell populations within the multivariate model.
Quantifying the markers for each phenotype involved in the multivariate model can provide additional insight into the differences between closely related cell populations.
For example, if %a computational multivariate model selects the %this is is an unnecessary restriction
 two phenotypes $1^+2^-3^+4^-5^+$ and $1^+2^-3^+4^-6^+$ are identified as correlates of a disease, and if markers $5$ and $6$ (which are the only differences between them) are the least important markers for the former and latter phenotypes respectively, then these two phenotypes are likely to correspond to the same cell population (as far as the correlation with the disease is concerned).
However, if markers $5$ and $6$ are the most important for the phenotypes, these can correspond to two biologically different cell populations.

To address these problems, we developed RchyOptimyx, a computational tool that uses dynamic programming and optimization techniques from graph theory to construct a cellular hierarchy, providing the best gating strategies to identify target populations to a desired level of purity or correlation with a clinical outcome, using the simplest possible combination of markers. 

%rewrote this next bit in the first two sentence to make it clearer, but this is already defined in the terms section and doesn't add anything at all by defining it here as well 
%A graph starts at level $0$ with the  phenotype that includes all cells (or the major sub group) and proceeds to the target population with $m$ markers at  level $m$ including $i$-marker phenotypes on the  $i^{th}$ level. 
%The phenotype with $0$ markers is the top most and the phenotype with  is a target cell population.

% the cell populations of interest. The cell population of interest doesn't have to be at the bottom, and there can be more than one. Combined your two definitions here (the one that was here, with the one that was in the methods. No need to define the same thing twice 2 different ways. 


% top-down diagram illustrating all cells on top most level and.
% These paths are then merged into a single cellular hierarchy  into a single that can provide a complete overview of the complex relationships between the markers involved.
% paragraph about visualization (including spade, gemstone, spice)
% use already available information from flowrepository and published materials to design panels

%A cellular hierarchy is a directed acyclic graph (DAG), embedded in a plane as a top-down diagram, with one node on the
%top most level representing all cells (or a major component therefore, such as T-cells) and nodes further down  showing
%more specific cell populations.  
%All the intermediate cell populations are placed in the hierarchy using parent-child
%relationships.
%The graph starts from level $0$ to level $m$ including $i$-marker phenotypes on $i^{th}$ level. 
%The phenotype with $0$ markers is the top most phenotype with all cells and the phenotype with $m$ markers is the cell population of interest.

\subsubsection{Materials and Methods}
Our methodology builds on the flowType pipeline\cite{Aghaeepour2012}.  flowType comprehensively  identifies  cell
populations defined by all possible gating strategies (hierarchies) in the data set using a partitioning strategy (\emph{e.g.}, clustering algorithm like flowMeans \cite{Aghaeepour2012}) and scores them by a statistical test (\emph{e.g.}, the log
rank test for difference in survival distributions).  
Given the list of all cell populations and their scores, RchyOptimyx uses a
dynamic programming approach to find the best cellular hierarchy within a reasonable time for interactive data analysis (\emph{e.g.} less than 2 minutes for 30 color data), as well as a number of best
suboptimal hierarchies, to enable mining of the space of best gating strategies and purities for a given target cell population.
% In this section, we describe how this is accomplished.


\subsubsection{Terms and Definitions}

Let $\mathcal{M}$ be the set of $m$ markers of interest (\emph{e.g.}, $\mathcal{M}=\{KI\mbox{--}67,CD28,CD45RO\}$), a
single marker phenotype be a phenotype having only one marker (\emph{e.g.}, $CD28^+$), a phenotype $P$ be a set of single
marker phenotypes (\emph{e.g.}, $P=KI\mbox{--}67^+CD28^-$), and $M$ (not to be mistaken with $\mathcal{M}$) be a phenotype of size $m$ that involves all of the
markers (\emph{e.g.} $M=KI\mbox{--}67^+CD28^-CD45RO^-$).  The power set of $M$, $\mathcal{P}(M)$, is of size $2^m$ and contains every possible subset of $M$.  
The scoring function $S(.)$ assigns a score to each member of $\mathcal{P}(M)$, such that higher values are assigned to more important phenotypes (\emph{e.g.}, those with a stronger correlation with a clinical outcome). 

Given an arbitrary $M$, the  directed acyclic  graph (DAG)  $G_M$ has $m + 1$ levels from $0$ to $m$, each level $i$ including every
member of $\mathcal{P}(M)$ of size $i$.  Node $s$ is connected to node $t$ with a directed edge $(s,t)$ if and only if
$|t|=|s|+1$ and the two associated sets of $s$ and $t$ differ only in one single phenotype marker (\emph{i.e.,} $t$ is
an immediate parent of $s$).  
Let the weight of edge $(s,t)$ be $-S(t)$ (so that paths with maximum score can be found by searching for paths with minimum total weight). 
%Nima: commented out the following sentence. They'll see this when they get there.
%The value $-S(t)$ is used in the dynamic programming and the $l$-minimum wight path finding algorithm, both of which discussed later. 


The node with $0$ markers is the root (or source) node, and the node with the complete set of markers is the sink node. 
A path from source to sink is called a hierarchy path, or simply a hierarchy. 
An example of graph $G_M$ for $M = KI\mbox{--}67^+CD4^-CCR5^+CD127^-$ is
illustrated in Supplementary Figure~\ref{r1:HIVClinicalOutcomeComplete}.

  \begin{landscape}
    \begin{figure}[ht]
      \begin{center}
        \includegraphics[width=13cm, angle=270]{figs/rchy/figs/HIVClinicalOutcomeComplete}
      \end{center}
      \caption{A complete cellular hierarchy for prediction of HIV's clinical outcome using KI67$^+$CD4$^-$CCR5$^+$CD127$^-$ T-cells. 
      The color of the nodes indicates the significance of the correlation with clinical outcome (p-value of the logrank test for the Cox proportional hazards model) and the width of each edge (arrow) shows the amount of change in this variable between the respective nodes.
        The positive and negative correlation of each immunophenotype with outcome can be seen from the arrow type leading to the node; however, as all correlations are negative in this hierarchy, only one arrow type is shown. 
      }
      \label{r1:HIVClinicalOutcomeComplete}
    \end{figure}
  \end{landscape}

The graph $G_M$ has $|\mathcal{P}(M)|=2^m$ nodes, one node for each parent phenotype of the phenotype of interest. 
The number of edges is equal to the number of markers ($m$), times the number of edges that have the specified marker.  
Each marker appears in $2^{m-1}$ nodes, therefore the number of edges is $m \times 2^{m-1}$. 

%Intuitively comparing two hierarchies, the one which goes through higher score nodes is better. Modeling this intuition
%on our constructed graph, considering the weight of a path as the score of each hierarchy satisfies reverse of the
%intuition, \emph{i.e.} better hierarchies get lower weight. Score of a hierarchy, thus, can be written as follows:

A scoring function is needed to find the best hierarchy.  This function should give a higher rank to hierarchies that
go through more important parent populations earlier (\emph{i.e.}, those that achieve a higher clinical significance with fewer
markers).  
Because each node of the hierarchy is a phenotype, and each phenotype has a given score value $S(.)$, we
use the \emph{total score} function $T(.)$ - the sum of all negated phenotype scores in the hierarchy - as the scoring
function:
\begin{equation}
\begin{split}
	T(\mathcal{H}) = \sum_{(s, t) \in E_{\mathcal{H}}}{W(s, t)} \\
	= \sum_{(s, t) \in E_{\mathcal{H}}}{-S(t)} \\
	= \sum_{t \in V_{\mathcal{H}} \setminus v_0}{-S(t)}
\end{split}
\end{equation}
where $\mathcal{H}$ is the given hierarchy, $E_{\mathcal{H}}$ is the set of edges of hierarchy $\mathcal{H}$, $V_{\mathcal{H}}$ is the
set of vertices of same hierarchy, and $v_0$ is the first node in the hierarchy.
Applying this function to $G_M$, the best hierarchy is the minimum weighted path in $G_M$. 
We note that, in principle, more complex functions can be used to compute the total score of a given hierarchy; for
example, in applications in which phenotypes with fewer markers are more important than the other phenotypes, an exponential 
function can be used to increase the weight of the earlier phenotypes in the hierarchy.


\subsubsection{Dynamic Programming to Identify the Best Hierarchy}
For cell populations characterized by $m$ markers, finding the best hierarchy 
by searching through all possible hierarchies would require time $O(m!)$, which is impractical for even moderately large $m$. 
To make this problem tractable using dynamic programming, we define \emph{best total score} function $T^*(.)$, which computes
the score of the best hierarchy leading to the given phenotype. $T^*(.)$ is defined recursively as follows: 

\begin{equation}
  T^*(P^k) = \left\{ \begin{array}{cl}
      -S(P^k) &\mbox{ if $k = 1$} \\
      \min\{T^*(P^k \setminus P^{k}_{i}) - S(P^k)|i=1,\dots,k\} &\mbox{ otherwise}
    \end{array} \right.,
  \label{totalscore}
\end{equation} 
where $P^k$ is a cell population defined by  $k$  single marker phenotypes, and $P^k \setminus P^k_{i}$ is $P^k$
with the $i^{th}$ single marker phenotype removed.  For example, if $P^3=KI\mbox{--}67^+CD28^-CD45RO^+$, then
$P^3\setminus P^3_1=CD28^-CD45RO^+$.  In other words, there is an edge from $P^k \setminus P^k_{i}$ to $P^k$ in $G_M$
where, $P^k$ is a subset of $M$. Also note that $-S(P^k)$ is the weight of the edge $(P^k, P^k \setminus P^k_{i})$ in
$G_M$.


%For each node in $G_M$, $T^*(.)$ calculates the score of the best hierarchy up to that node. 
Using dynamic programming, we calculate the value of $T^*(.)$, iterating from level $0$ to $m$ on $G_M$.
Calculating each node's score requires a number of constant time operations equal to the number of edges entering the
node.  Therefore, the total number of operations is proportional to total number of edges ($m \times 2^{m-1}$), and the
overall time complexity of our programming procedure for determining $T^*(.)$ values for all phenotypes in the graph is
$O(m \times 2^{m-1})$.
An illustration of the dynamic programming space for three dimensional space, \emph{i.e.} having three markers, as well as 
two paths in that space is shown in Figure \ref{r1:3dDP}.


%\begin{landscape}
\begin{figure}[ht]
  \begin{center}
    \includegraphics[width=\textwidth]{figs/rchy/2path}
  \end{center}
  \caption{Dynamic programming algorithm for two cell populations defined by $3$ markers.
    The best path for each of the cell population is shown in red and blue respectively.
    As an example, the red path ends at CD4$^+$CCR5$^+$CD127$^+$.
    Three markers are available to be added.
    First, CD4 is added (changes from does not matter to positive).
    Then two options will be available for the next step (CD127 and CCR5).
    After selection of CCR5, only one option will be left for the final step (CD127).
    Therefore for three markers, $\frac{3 \cdot (3-1)}{2}=6$ comparisons were required.
	\textbf{Left:} A hierarchy for the two paths. The label of an edge is the name of the single marker phenotype that
	is the difference between its head set ($s$) and its tail set ($t$).  \textbf{Right:} the dynamic programming space
	for the $3$ markers. Black spheres mark the nodes in the dynamic programing space used by the two paths. 
	The colors of the nodes on the left match that of the square tori on the right and  correspond to the relative score of each cell population. 
}
  \label{r1:3dDP}
\end{figure}
%\end{landscape}

%To implement this computationally, we represented cell populations as base $3$ numbers.  For example, given the set of
%markers \{A, B, C, and D\}, the A$^-$B$^+$D$^+$ population can be represented as $1202$, where $0$, $1$, and $2$
%represent \emph{not considered}, \emph{negative}, and \emph{positive}, respectively \cite{janet8exhaustive,
%  Aghaeepour2011Early}. Dynamic programming space will include one node for each possible phenotype and the phenotype is
%associated with the corresponding node. Therefore the dynamic programming space is of size $3^m$, where $m$ is the
%number of markers \cite{janet8exhaustive}.  Given a node in the dynamic programming space (\emph{i.e.,} the population
%of interest), the best path (or cellular hierarchy) to the base node (\emph{i.e.,} all cells) can be calculated in
%$m+(m-1)+...+1=\frac{m\cdot (m+1)}{2} \in O(m^2)$, (Figure \ref{r1:3dDP}).  Therefore the total time complexity of the
%dynamic approach is $O(3^m)$. % (the time required for generating the dynamic table of size $3^m$).

\subsubsection{Search for Near-Optimal Hierarchies}
The hierarchy selected by the dynamic programming algorithm is the best gating strategy for a given cell population.
%%NA: Changed awkward wording
However, we would also like to identify alternate gating strategies with slightly less desirable scores.  
To find these near-optimal paths, we reformulate the problem as identification of a desired number of minimum weight paths: %%NA Shouldn't break paragraph here
In $G_M$, the minimum weight path from source to sink is the best hierarchy (identical to the one generated by dynamic programming).  
To generate additional, sub-optimal hierarchies, a list of the next minimum weight paths must also be generated. 
These paths can be identified using the method by Eppstein  \cite{eppstein1998finding}.  
As noted in the original article, elaborating the details of this algorithm is complicated and requires substantial
background in algorithm design, which is well beyond the scope of this work.  Briefly, this method uses the minimum
spanning tree of $G_M$ and computes a heap structure for each node; it then merges the heaps in an efficient way to
construct a 4-heap data structure. Using this 4-heap and a given arbitrary number $l$ (the number of desired paths), it generates $l$-minimum weight paths in time $O(e + v + l)$ for a DAG with $e$ edges and $v$ nodes (see Theorem $4$ of \cite{eppstein1998finding} for details).


Hence, the time complexity of our algorithm can be calculated based on the number of
edges and nodes using the time complexity of the $l$-minimum weight paths method:
\begin{equation}
  \begin{split}
    O(e + v + l) & = O(m \times 2^{m-1} + 2^{m} + l)\\
    & = O(m \times 2^{m-1} + 2 \times 2^{m-1} + l)\\
    & = O((m+2) \times 2^{m-1} + l).\\
  \end{split}
  \label{order_equation}
\end{equation}

For example, the number of operations with our approach on a dataset with $m=10$ markers would be $\approx 10^4$
compared to $\approx 3 \times 10^{6}$ for the exhaustive search approach.  
Our method therefore takes
$\approx 0.23$ CPU seconds vs $\approx 69$ CPU seconds for exhaustive search, run under 64 bit Linux (version 3.3) on $2.93GHz$
Intel Xeon CPU with sufficient memory (proportional to $2^M$). 
For a phenotype involving $m=20$ markers, these numbers increase to $\approx 1.2$ CPU seconds vs $\approx 10^{11}$ CPU
seconds (more than $4000$ years), respectively.  Even for a phenotype involving $m=30$ markers measured by a CyTOF assay
(mass spectrometry-flow cytometry hybrid device \cite{ornatsky2010highly,bendall2011single,Chattopadhyay2012Cytometry}),
RchyOptimyx remains feasible, with a runtime of $\approx 102$ CPU seconds, while the brute-force method would take
$\approx 10^{22}$ CPU seconds.  The final output of RchyOptimyx is the corresponding subgraph of $G_M$ that includes all
calculated paths (\emph{i.e.}, the optimized hierarchy, \emph{e.g.}, Fig.~\ref{r1:HIVClinicalOutcome}).

  \begin{figure}[ht]
    \begin{center}
      \includegraphics[width=.7\textwidth]{figs/rchy/figs/HIVClinicalOutcome}
    \end{center}
    \caption{An optimized cellular hierarchy for prediction of HIV's clinical outcome using KI67$^+$CD4$^-$CCR5$^+$CD127$^-$ T-cells. 
The color of the nodes shows the significance of the correlation with clinical outcome (p-value of the logrank test for the Cox proportional hazards model) and the width of each edge (arrow) shows the amount of change in this variable between the respective nodes. 
}
    \label{r1:HIVClinicalOutcome}
  \end{figure}


%\tin{don't have single sentence paragraphs}

% \subsection*{Trimming the Hierarchies}
% In some cases, markers are not only redundant, but also have a negative impact on the cell population (the parent population can be better than the child).
% In these cases, the lower parts of the hierarchy do not provide any additional information and can be removed.
% I don't think this is worth mentioning. If it is, it should be incorporated into the above subsection, not a subsection on its own.

% For example, the original hierarchy for identification of KI-67$^+$ cells is longer than the one demonstrated in Figure \ref{KI67Overlap} but the additional markers were not useful in increasing the overall accuracy.

\subsubsection{Datasets}

We validated RchyOptimyx on two high-dimensional datasets, produced by mass and polychromatic flow cytometry.

\emph{Mass cytometry analysis of bone marrow cells from normal donors}
In this dataset, $31$ parameters were measured for mononuclear cells from a healthy human bone marrow (see  \cite{bendall2011single} for details).
We used the results of three assays on samples subject to \emph{ex vivo} stimulation by IL7 (measured by pSTAT5), BCR (measured by pBLNK), and LPS (measured by p-p38) as well as an unstimulated control.
$13$ surface markers were included in the analysis: CD3, CD45, CD45RA, CD19, CD11b, CD4, CD8, CD20, CD34, CD33, CD123, CD38, and CD90.
Singlets were gated manually, as described in the original publication.



\emph{Polychromatic flow cytometry analysis of HIV$^+$ patients}
This dataset consists of $13$ color PFC assays of $466$ HIV$^+$ subjects enrolled in the Infectious Disease Clinical Research Program's HIV Natural History Study.
Basic demographic characteristics of this dataset are described elsewhere \cite{weintrob2008increasing}.
Cryopreserved peripheral blood mononuclear cells stored within 18 months of the date of seroconversion were analyzed using PFC as described by Ganesan \emph{et al.} \cite{Ganesan2010Immunologic}. 
The cohort included 135 death/AIDS events, as defined by 1993 guidelines \cite{castro1992revised}.
The date of the last follow-up or initiation of highly active anti-retroviral therapy (HAART) was considered a censoring event.
CD14 and V-amine dye were used to exclude monocytes and dead cells, respectively, CD3 was used to gate T-cells.
Using the staining panel and flowType, we enumerated various subsets of naive and memory T-cells, defined by CD4, CD8, CD45RO, CD27, CD28, CD57, CCR5, CCR7, CD127, and KI-67.
Using a log rank test with Bonferroni's multiple test correction, we scored each subset (cell population) in terms of its correlation with HIV progression \cite{Aghaeepour2012}.

\subsubsection*{Results}

% To validate RchyOptimyx, we applied it to three proof-of-concept examples in which cellular hierarchies for a given population of interest were generated and optimized using an external variable (clinical outcome, overlap with naive T-cells, and overlap with KI-67$^+$ T-cells without using KI-67).

\subsubsection{Designing a Panel to Detect a Population Expressing an Intracellular Marker using Surface Markers}
In this use--case, our goal was to identify cell populations that are affected by different stimulations in the mass cytometry dataset.
We used flowType to identify a list of populations that had a high overlap with either the IL3$^+$, BCR$^+$, or LPS$^+$ populations (determined manually - see Fig.~\ref{r1:overlaps}).
For each cell population, this value was calculated as the difference in its intersection with the IL3$^+$, BCR$^+$, or LPS$^+$ compartments between the stimulated and unstimulated sample.
For example, for a given cell population CP, the overlap with IL3$^+$ was defined as:
\begin{equation}\small
  Overlap^{IL3^+}(CP)=\left(\frac{\# \; \mbox{IL3}^+cells \; in \; CP}{\# \; cells \; in \; CP}\right)_{stim}-  \left(\frac{\# \; \mbox{IL3}^+cells \; in \; CP}{\# \; cells \; in \; CP}\right)_{unstim}
\end{equation}

\begin{figure}[!ht]
  \begin{center}
    \includegraphics[width=\textwidth]{figs/rchy/figs/overlaps}
  \end{center}
  \caption{All immunophenotypes ordered by their overlap with the cell population of interest. The red dashed lines indicate the cutoffs used for selecting the immunophenotypes with ``high overlap''.}
  \label{r1:overlaps}
\end{figure}


The immunophenotypes with a high overlap, as identified by flowType, are listed in Tables~\ref{apx:BCR},~\ref{apx:IL7}, and~\ref{apx:LPS}.
These immunophenotypes were analyzed using RchyOptimyx (\emph{e.g.}, Fig.~\ref{r1:CyTOF} for BCR) and then merged into a single graph, shown in Fig.~\ref{r1:CyTOF}.
This graph suggests that T-cells (CD3$^+$) followed by cytotoxic T-cells (CD3$^+$CD4$^+$) are the main parent populations that are affected by IL7 stimulation (panel A).
As expected, BCR stimulation affected B-cells (CD19$^+$CD20$^+$CD3$^-$), and LPS stimulation increased the proportion of CD19$^-$CD33$^+$CD3$^-$ cells (Panels B and C, respectively).
These results are generally consistent with those reported in the original study (Figure 2 and panel C of Figure 3 of \cite{bendall2011single}).

%\begin{landscape}
\begin{figure}[!h]
  \begin{center}
    \begin{tabular}{c||c||c}
      \includegraphics[width=.6\textwidth,angle=270]{figs/rchy/IL7}&
      \includegraphics[width=.6\textwidth,angle=270]{figs/rchy/BCR}&
      \includegraphics[width=.6\textwidth,angle=270]{figs/rchy/LPS}\\
      (A) IL7/pSTAT5 &(B) BCR/pBLNK &(C) LPS/p-p38\\
    \end{tabular}
  \end{center}
  \caption{Three optimized hierarchies for identification of cell populations with maximum response to IL7, BCR, and LPS measured by pSTAT5, pBLNK, and p-p38, respectively.
    The colour of the nodes and the thickness of the edges shows the proportion and change in proportion of cells expressing the intracellular marker of interest, respectively.
  }
  \label{r1:CyTOF}
\end{figure}
%\end{landscape}


\subsubsection{Simplifying Gating Strategies}

Here we use RchyOptimyx to demonstrate  an example of the use case of establishing  a simpler combination of markers that can be used to  identify a target population at a desired level of purity. 
For analysis of the PFC dataset, Ganesan \emph{et al.} used a strict, but potentially redundant definition for naive T-cells, of CD28$^+$CD45RO$^-$CD57$^-$CCR5$^-$CD27$^+$CCR7$^+$, within the CD3$^+$CD14$^-$ compartment \cite{Ganesan2010Immunologic}.
The purity of a given parent cell population (CP) of  this target was defined as its mean purity for the strictly-defined naive T-cells:
\begin{equation}\small
  Purity(CP)=\frac{\sum_{All \; Samples}\frac{\# \; CD28^+ CD45RO^- CD57^- CCR5^- CD27^+ CCR7^+ \; cells}{\# \; cells \; in \; CP}}{\# \; Samples}
\end{equation}

Figure \ref{r1:NaiveOverlap} shows the results of analysis with RchyOptimyx where a combination of only three markers (CD45RO$^-$CCR5$^-$CCR7$^+$) identified the strict naive T cell population to $95\%$ purity (within the CD3$^+$CD14$^-$ compartment).
The range of available purities, and determination of an appropriate cutoff is experiment dependent (\emph{e.g.}, on the range of available markers or biological question being researched) and this result is only provided as an example of the utility. 

\begin{figure}[ht]
  \begin{center}
    \includegraphics[width=.7\textwidth]{figs/rchy/NaiveOverlap}
  \end{center}
  \caption{An optimized cellular hierarchy for identifying naive T-cells. The color of the nodes and the thickness of the edges shows the purity and change in purity of the original naive phenotype within the given cell population, respectively.
  }
  \label{r1:NaiveOverlap}
\end{figure}


\subsubsection{Characterization of a Large Number of Immunophenotypes}
Here we use RchyOptimyx to demonstrate an example of the use-case of summarizing a large list of immunophenotypes of interest (as identified by a bioinformatics pipeline) into a single hierarchy using their most important common parent populations.

In a previous study of the PFC dataset, we identified $101$ immunophenotypes (Table~\ref{apx:tsl}) in HIV$^+$ patients that had a statistically significant correlation with HIV's progression \cite{Aghaeepour2012}.
%We used RchyOptimyx to simplify these cell populations and then merge them into a single graph. %duplicated line 306
The score of each population was calculated as $-log_{10}(p)$ where $p$ was the p-value of the logrank test before adjustment for multiple testing (higher values represent a stronger correlation with the clinical outcome).
The $101$ immunophenotypes were analyzed using RchyOptimyx and the resulting hierarchies were merged into a single graph (Figure \ref{r1:merged}).
This graph indicated  three groups of immunophenotypes that were significantly correlated with HIV's outcome (left, center, and right branches).
The left branch consisted of KI-67$^+$CD4$^-$CCR5$^+$CD127$^-$ T-cells.
These cells were thought to be statistical significant mainly because they are long-lived (CD127$^-$) T-cells with high proliferation (KI-67$^+$). 
RchyOptimyx showed that the significance of this population is related to the KI-67$^+$CCR5$^+$ compartment and not CD127$^-$ (Figure \ref{r1:merged}, the left branch) as the CD127 marker is not needed to achieve the approximately the same score.  
This is in agreement with the results of two recent studies \cite{gordon2010disruption,jaspan2011immune}.
The terminal node of the center branch consisted of seven markers (CD45RO$^-$CD8$^+$CD57$^+$CCR5$^-$CD27$^+$CCR7$^-$CD127$^-$). 
RchyOptimyx revealed that its most important parent population is CD8$^+$CCR7$^-$CD127$^-$, with a weaker correlation with the clinical outcome.
Finally, the right branch (CD28$^-$CD45RO$^+$CD4$^-$CD57$^-$ CD27$^-$CD127$^-$) suggests several parent populations with minimal overlap and strong correlation with the clinical outcome (\emph{e.g.,} CD28$^-$CD4$^-$CD57$^-$CD127$^-$ and CD45RO$^+$ CD4$^-$CD127$^-$). 

\begin{landscape}
  \begin{figure}[!hp]
    \begin{center}
      \includegraphics[width=\textwidth,angle=270]{figs/rchy/merged}
    \end{center}
    \caption{An optimized hierarchy for all three populations correlated with protection against HIV.
      The color of the nodes shows the significance of the correlation with the clinical outcome (p-value of the logrank test for the Cox proportional hazards model) and the width of each edge (arrow) shows the amount of change in this variable between the respective nodes.
      The positive and negative correlation of each immunophenotype with outcome can be seen from the arrow type leading to the node; however as all correlations are negative in this hierarchy, only one arrow type is shown. 
   }
    \label{r1:merged}
  \end{figure}
\end{landscape}


\subsubsection{Discussion}
Sequential analysis of the markers involved in manual or automated identification of cell populations is fundamental to our understanding of the characteristics of the cell population.
In sequential gating, the order in which the gates have been applied does not affect the final results.
However, ordering the gates by their relative importance has two use-cases: 1) identifying a cell population of interest, using the smallest possible panel of markers; 2) summarizing a long list of closely related (and perhaps overlapping) immunophenotypes by identifying their most important common parent populations.
%For immunophenotypes with fewer markers, the importance of the markers can be manually investigated (see example of a complete cellular hierarchy for a phenotype with four markers in Figure S3).
%By visual inspection one can find that KI-67 and CCR5 are the two most important markers.
However, increasing the number of markers quickly renders this approach infeasible (\emph{e.g.}, Fig.~\ref{r1:NaiveOverlapComplete} for only six markers).

%\begin{landscape}
\begin{figure}[!ht]
  \begin{center}
    \includegraphics[width=.8\textwidth, angle=270]{figs/rchy/figs/NaiveOverlapComplete}
  \end{center}
  \caption{A complete cellular hierarchy for identifying naive T-cells. The colour of the nodes and the thickness of the edges have been removed to facilitate visualization of the complex graph.
  }
  \label{r1:NaiveOverlapComplete}
\end{figure}
%\end{landscape}

To address this challenge, we  developed RchyOptimyx, a computational tool that automatically characterizes the complex findings of high-dimensional exploratory FCM studies.
RchyOptimyx sorts all parent populations of an immunophenotype of interest into hierarchies, and selects those hierarchies that are better able to maintain the characteristics of the immunophenotype of interest (\emph{e.g.,} correlation with a clinical outcome).
This reveals the best order in which markers can be excluded from an immunophenotype.
RchyOptimyx uses dynamic programming and efficient tools from graph theory to make the problem tractable using the computing resources readily available in most laboratories. 

Since most cells can be described using more than one combination of markers, there usually are several alternative cellular hierarchies associated with every population.
RchyOptimyx finds all these ``paths'' and merges them into a single hierarchy, starting from ``all cells'', or any arbitrary point in a hierarchy,  and finishing at the terminal population of interest.
This reveals the relationships between different gating strategies and how they differ, and also facilitates the reproduction of high--dimensional exploratory studies using low--color instruments.
The ability to suggest multiple panels is particularly important when designing new panels, because the choice of markers depends on a large number of external parameters including, but not limited to, reagents available through vendors, potential spectral overlaps, the instruments available, and budget limitations.


Another important use-case for RchyOptimyx is in the  interpretation of the findings of bioinformatics pipelines.
While these pipelines have recently been very successful in identifying cell populations correlated with clinical outcomes, their findings cannot be easily understood for two reasons: 1) they usually rely on high-dimensional clustering of the data and therefore cannot propose gating strategies for reproduction of their results; 2) their predictive power often relies on a large list of immunophenotypes. 
Some of these immunophenotypes are closely related (\emph{e.g.}, refer to close or overlapping cell populations) while others are not. 
RchyOptimyx addresses the first problem by suggesting optimized gating hierarchies for identification of these cell populations to a desired level of purity or correlation with clinical outcome. 
The latter problem is addressed by summarizing closely related immunophenotypes using their most important common parents.

%However, increasing the number of calculated hierarchies can make the visualizations complicated, so RchyOptimyx allows the user to manually control trade-off for each specific use-case.

In evaluating  RchyOptimyx, we  combined its functionality with the
automated gating functionality provided by flowMeans and flowType. 
However,  RchyOptimyx can be built upon the results of any cell  %is the grammar here correct? RchyOptimyx can be built upon...?
population identification method, including manual analysis, provided
all intermediate cell populations (\emph{i.e.,} each layer, removing one marker at a time) from the cell population of interest  up to the desired start of the hierarchy are provided to the method.

We evaluated RchyOptimyx for three use-cases, using a small but high-dimensional mass cytometry dataset and a clinical dataset of high-dimensional conventional FCM  assays of $466$ patients, previously analyzed by both manual and automated analysis.
First, we constructed cellular hierarchies for identification of cells that were produced in response to different stimulations.
This use-case represents the problem of designing panels of surface markers (primarily for sorting) for cells that can only be defined using their intra-cellular signature (possibly after proper stimulation). 
For example, plasmacytoid dendritic cell (PDC)s are known to express the toll-like receptor 9 (TLR9) in response to stimulation using CpG \cite{krug2001toll}. 
A large number of surface candidates were recently proposed for PDCs \cite{marafioti2008novel,swiecki2010unraveling,schuster2010co,cao2009molecular}.
An interesting direction to extend this work  would be to measure all these markers in a single panel, subject to CpG stimulation (using appropriate controls) to design a panel of surface markers for PDCs.
In this  case, TLR9 could be used as the external variable for optimization.

Second, we demonstrated that RchyOptimyx can be used to simplify existing gating strategies, using as an example the identification of naive T-cells previously defined using a complex panel of six markers to a $95\%$ purity using only three. 
This proof-of-concept use-case is relevant when a subset of markers needs to be selected for reproduction of the results using fewer colors.
For certain biological use--cases, purity of higher than $95\%$ can be required.
For such use--cases, a larger number of markers for exclusion of non-naive T-cells should be included in the panel.

Third, we showed that RchyOptimyx, together with a complex bioinformatics pipeline, can analyze a large high-dimensional clinical dataset, to reveal correlates of a clinical outcome, hidden from previous manual and automated analysis of the same dataset.
In addition, RchyOptimyx suggests the best gating strategies and marker panels for reproduction of these results in low-color settings.
By identifying the best cellular hierarchies, RchyOptimyx allows the user to make an informed decision about the trade-off between the number of markers and the significance of the correlation with the clinical outcome.
This feature is particularly important in hypothesis generating studies that need to be further validated using large clinical studies.
%\tin{In the discussion, highlight why this was missed with the earlier approach, and how the hierarchical approach now allows us to identify this, also highlight that manual analysis failed as well. Mention how this is hypothesis generating for exploratory research?} 

For the third example, it is important to note that the correct measure for the amount of correlation with a clinical outcome is an effect size (such as the root squared error of the estimated proportional hazard). 
However, such effect size does not provide any information about the significance of the correlation. 
As RchyOptimyx is intended to be a decision support tool, and in this case the decision is the degree to which a cell population can be generalized while maintaining the statistical significance of the correlation, we decided that the p-values of the log-rank tests were more appropriate for optimization of the hierarchies.
To support this decision, we empirically investigated the differences between the p-values and effect sizes of the Cox proportional hazard models (Fig.~\ref{r1:effectsizevspv}) and concluded that these values are highly correlated (which is not surprising considering the large size of our cohort).
It should be noted that as RchyOptimyx allows the user to choose which measure to provide, they can make this decision as appropriate for their specific data.

\begin{figure}[!ht]
  \begin{center}
    \includegraphics[width=10cm]{figs/rchy/figs/cor}
  \end{center}
  \caption{The correlation between the effect sizes and p-values of the log rank tests for the Cox proportional hazards models for each immunophenotype. The Pearson correlation coefficient was determined as $0.997$, indicating a highly significant correlation with a p-value  $< 2.2 \times 10^{-16}$.}
  \label{r1:effectsizevspv}
\end{figure}



% \hide{
%   RchyOptimyx, when applied to high-dimensional datasets \emph{e.g.,} through FlowRepository), can provide significant guidance into panel design.
%   However, it is important to note that in practice, many other factors including the instruments available, the conjugated antibodies produced by vendors, and potential spectral overlaps, should be considered as well.}



% RchyOptimyx builds on the RchyOptimyx uses this information and then adds the structure of the hierarchy based on parent-child relationships between the cell populations and can be used to study a set of one or more cell populations based on an external score, often measured across a large cohort. Nima: This is not true. flowType doesn't require flowMeans. 
% yes, but I think I made this point now more clear at top of discussion
% These parent-child relationships enable RchyOptimyx to reveal the importance of each marker for maximizing the external score and suggest the best parent population for identification of the cell population of interest.
% In other words, SPADE helps in identifying the cell population of interest (like several other bioinformatics pipelines and manual gating as reviewed in the introduction).
% The identified cell population of interest can then be further analyzed using RchyOptimyx to reveal the importance of the markers and various gating hierarchies.
% \tin{plus we get significance, completely automated} RchyOptimyx doesn't have anything to do with statistical significance. flowType does. If you want to compare SPADE against anything, it has to be flowType (both are phenotyping algorithms). The cool thing about flowType/RchyOptimyx (as I have tried to explain) is that it models the data using parent/child relationships, which allows it to comment on which parent best describes the cell population of interest.
% \tin{ so what is new? why should we care if you just get the same answer? Address this in discussion}

The concept of computationally extracting cellular hierarchies from FCM data has previously been introduced by the SPADE algorithm \cite{bendall2011single, qiu2011extracting}.
SPADE generates a large number of multidimensional clusters and then connects them to each other using the distance between their mean/median fluorescence intensities. 
These are then manually annotated by biologists with domain knowledge.
This makes SPADE useful for identification and visualization of a large number of clusters, particularly when expression of markers change gradually (\emph{e.g.,} cell-cycle analysis and some intracellular studies).
However, the hierarchies generated by SPADE are logically and conceptually different from those generated by RchyOptimyx and have different use-cases.
For example, the results of the mass cytometry dataset presented here are very close to results previously obtained from SPADE analysis.
However, SPADE required manual annotation of the results by a human expert, using different plots demonstrating the expression of different surface markers and the intra--cellular marker of interest (Figure 2 and panel C of Figure 3 of \cite{bendall2011single}).
More complicated relationships that involve several markers cannot be easily identified by these manual annotations.
In addition, SPADE is limited in that the relationships between cell populations is exclusively defined using the multidimensional distances between them.
However, two cell populations that are close to each other in the multidimensional space can be far in terms of specific markers (which can be the most important ones).
%Therefore, two cell populations can be close and far from each other at the same time, based on different parent populations.
The cellular hierarchies generated by RchyOptimyx are based on parent-child relationships, guided by an external variable (cell populations that have common parents with similar patterns of correlation with a clinical outcome or intracellular response to stimulation are grouped together).
%In RchyOptimyx, the final similarity of two cell populations is determined after breaking them down to single marker gates, measuring the importance of each gate using an external variable, and identifying their common parents.
%In RchyOptimyx, two cell populations are close to each other if they have common parents with similar patterns of correlation with a clinical outcome or response to stimulation (in contrast to other methods in which similarity is defined based on MFI).
This enables RchyOptimyx to automatically annotate a large number of cell populations identified by other methods (\emph{e.g.}, manual gating or SPADE) in terms of the importance of the markers involved and summarize them in a single hierarchy.


There are several directions in which this work can be extended.
RchyOptimyx provides no information about the robustness of the hierarchies.
Bootstrapping strategies could be used to produce confidence intervals for the tree structure and increase generalizability to previously unseen data \cite{suzuki2006pvclust}.
Also, our current implementation of RchyOptimyx assumes that every marker can be partitioned into a positive and negative population.
While the underlying theory does support additional (\emph{e.g.}, dim, bright, or low) populations, parts of the software package would need to be modified to accommodate these cases.


%\hide{
%\newpage
%\section*{Conclusions}
%RchyOptimyx is a methodology for extracting cellular hierarchies that can characterize complex cell populations in terms of the markers involved, visualize them, and guide the design of new marker panel and gating strategies.
%}
\subsubsection{Availability}
The RchyOptimyx R package (including source code, documentation, and examples) is freely available under an open source license (\emph{Artistic 2.0}) and can be obtained from Bioconductor.
The raw data and meta-data used in this study is publicly available through FlowRepository.org (under experiment ID \emph{FR-FCM-ZZZK}) and through Cytobank.org (under experiment ID \emph{6033}) for the PFC and CyTOF datasets, respectively.





%\hide{
%  \begin{figure}[ht]
%    \begin{center}
%      \includegraphics[width=13cm, angle=270]{figs/rchy/KI67Overlap}
%    \end{center}
%    \caption{A cellular hierarchy for identifying KI-67$^+$ T-cells using surface markers. The color of the nodes and the thickness of the edges shows the proportion and change in proportion of KI-67$^+$ T-cells, respectively.}
%    \label{r1:KI67Overlap}
%  \end{figure}
%}

%  \begin{figure}[ht]
%    \begin{center}
%      \includegraphics[height=9cm, angle=0]{figs/rchy/figs/cdf}
%    \end{center}
%    \caption{Empirical cumulative distribution function of the $24$ possible paths (one dot per path, with some dots overlapping) for constructing the KI67$^+$CD4$^-$CCR5$^+$CD127$^-$ population. A change point in the path scores was defined manually (the red dashed line). The paths above the dashed line are those selected for construction of Figure \ref{r1:HIVClinicalOutcome}.}
%    \label{r1:cdf}
%  \end{figure}


%  \begin{landscape}
%  \begin{figure}[!ht]
%    \begin{center}
%      \begin{tabular}{c||c||c}
%        \includegraphics[width=4in,angle=270]{figs/rchy/figs/IL7Proportions}&
%        \includegraphics[width=4in,angle=270]{figs/rchy/figs/BCRProportions}&
%        \includegraphics[width=4in,angle=270]{figs/rchy/figs/LPSProportions}\\
%        (A) IL7 &(B) BCR &(C) LPS\\
%      \end{tabular}
%    \end{center}
%    \caption{Average F-measures of all pairs of results for the cell populations across all samples in the HSCT dataset (\emph{i.e.,} one heatmap for every cell population in the reference).}
%    \label{CyTOFProp}
%  \end{figure}
%  \end{landscape}

  \begin{figure}[!ht]
    \begin{center}
      \includegraphics[width=\textwidth, angle=270]{figs/rchy/figs/PropNaiveOverlap}
    \end{center}
    \caption{An optimized cellular hierarchy for identifying naive T-cells. The colour of the nodes and the thickness of the edges shows the purity and change in purity of the original naive phenotype within the given cell population, respectively.
      This is similar to Figure 6 in the main text except the color of the border of the nodes shows the cell proportion of the cell population.
    }
    \label{r1:NaiveOverlapProp}
  \end{figure}

 \begin{figure}[!ht]
    \begin{center}
      \includegraphics[width=.7\textwidth, angle=270]{figs/rchy/figs/PropKI67Overlap}
    \end{center}
    \caption{A cellular hierarchy for identifying KI-67$^+$ T-cells using surface markers. The colour of the nodes and the thickness of the edges shows the proportion and change in proportion of KI-67$^+$ T-cells, respectively.
      This is similar to Figure 7 in the main text except the color of the border of the nodes shows the cell proportion of the cell population.}
    \label{r1:KI67Overlap}
  \end{figure}

 \subsection{flowType/RchyOptimyx pipiline}
 Flow cytometry has undergone a ``chromatic explosion'' over the past decade and can now measure 17 markers at once for each of hundreds of thousands of individual cells \cite{Chattopadhyay2008}.
Since then, mass cytometry has enabled measurement of 30--45 markers per cell \cite{Bendall2012a}, while single-cell multiplexed RT-qPCR can measure 50--96 mRNAs per cell \cite{White2011}. 
The growth in high-throughput single-cell data continues to outpace development of corresponding bioinformatics techniques \cite{Chattopadhyay2008}.
To answer this challenge, we previously developed flowType \cite{Aghaeepour2012} and RchyOptimyx \cite{Aghaeepour2012a}. 
flowType uses partitioning of cells, either manually or by clustering, into positive or negative for each marker to enumerate all cell types in a sample, e.g. \cite{Aghaeepour2013Critical}.
RchyOptimyx measures the importance of these cell types by correlating their abundance to external outcomes, such as disease state or patient survival, and distills the identified phenotypes to their simplest possible form. 
These packages have been used to identify several novel cell populations correlated with HIV outcome \cite{Aghaeepour2012}. 
More recently, this pipeline has been used to evaluate standardised immunological panels \cite{Villanovaa2013Computational}, to optimise lymphoma diagnosis \cite{Craig2013Computational}, and to analyse a range of other clinical data (unpublished).

However, the higher dimensionality of data produced by mass cytometry generates up to $3^{45}\approx10^{21}$ possible cell types, with an even greater number (up to $3^{96}\approx10^{45}$) for single-cell qPCR; these magnitudes are beyond the capabilities of flowType and RchyOptimyx.
Furthermore, flowType and RchyOptimyx have thus far only treated cells as being either positive or negative for a marker.
In practice, many biomarkers can have a range of expression levels such as ``dim'' and ``bright''. 
In this application note, we detail architectural improvements to flowType and RchyOptimyx to overcome these limitations.  
\subsubsection{Approach}
Our primary challenge was to enable flowType to generate a number of cell types tractable on most common workstations (e.g. those with 4--12GB of RAM).
%RB\textbf{define tractable in terms of some kind of upper/lower bounds, goals need to be specific and measurable, you do this two sentences down, but add here}
We hereafter denote the original flowType implementation as flowType-BF (brute force), and the new version as flowType-DP (dynamic programming).
Whereas flowType-BF completely enumerates all cell types over all $[1,...,m]$ markers, we opted  in flowType-DP to use a breadth-first strategy of enumerating all cell types defined over a subset of $k\leq m$ markers. %Holger had a comment on how k is chosen for a given m. This is very much subjective; anecdotally we have noticed that 6 is a good value, as any more markers than that tend to converge down to 6 in RchyOptimyx analysis. However, any value up until memory runs out is possible. This is actually a limitation in mass cytometry or single cell RT-qPCR data, where 4 is the upper limit possible within memory.
%Solution: use the memory estimation code
We provide a memory use estimation function, to assist users in finding a $k$ that fits within the limits of their hardware.
To improve computation time, in flowType-DP we implemented a dynamic programming approach, which exploits the fact that cell types can be arranged into a hierarchy, and membership of any given cell type over $n$ markers is equal to the intersection of one of its parent types (over $n-1$ markers) with a single-marker cell type.
flowType-DP first enumerates all cell types involving only 1 marker by simple partitioning and then iterates over $2,...,k$ markers, computing all cell types for each level $n$ by set intersections between corresponding cell types in levels $n-1$ and $1$.

For example, membership of the cell type CD45$^{++}$CD117$^+$CD34$^-$ is computed as follows:
\begin{align}
  &\lbrace \text{CD45}^{++}\text{CD117}^+\text{CD34}^-\rbrace & \nonumber
  \\= &\lbrace \text{CD45}^{++}\text{CD117}^+ \rbrace \cap \lbrace \text{CD34}^- \rbrace & \nonumber
  \\= &\lbrace \text{CD45}^{++} \rbrace \cap \lbrace \text{CD117}^+ \rbrace \cap \lbrace \text{CD34}^- \rbrace &
\end{align}

% In addition, we implemented flowType-DP in C++, using the Boost library dynamic\_bitset class to represent cell type membership sets.

To allow partitioning into levels other than positive and negative, we used a string representation for cell types.
The string has one integer character for every marker, denoting the partition, or zero if the marker is not used.
Values $1 , ..., n$ denote partitions $1$ to $n$.
For example, if the set of markers were $\lbrace$CD3, CD45, CD13, CD117, CD34$\rbrace$ the cell type CD45$^{++}$CD117$^+$CD34$^-$ would be represented by \verb#03021#. 
RchyOptimyx uses a dynamic programing algorithm for efficiently constructing $k$-shortest paths \cite{eppstein1998finding}. 
We modified RchyOptimyx' graph construction component to be able to handle more than one partition per marker.

\subsubsection{Results and Discussion}
We evaluated flowType-DP against flowType-BF on a 10-marker dataset available from Flow Repository (ID FR-FCM-ZZZK) \cite{Aghaeepour2012}.
flowType-DP showed a substantial speedup over flowType-BF, which increases exponentially with the number of cells and markers. 
For example, at $10^6$ cells and $10$ markers, flowType-DP is 14 times faster (see Fig. \ref{fig:01}a and b). %Change to e.g. 14-fold at 100k cells; fold-difference grows exponentially
Comparison on larger datasets was not possible, due to the limitations of flowType-BF.

\begin{figure*}[!ht]%figure1
\center{\includegraphics[width=\textwidth]{figs/flowtype2-figure}}
\caption{
\textbf{a-b. Run time comparison of flowType-DP to flowType-BF} in terms of number of cells (a) and number of markers (b). 
\textbf{c-d. Possible thresholds for marker combinations using flowType-DP} for typical mass cytometry data (c) and polychromatic flow cytometry data (d). 
\textbf{e-f. Three/four-partition flowType-generated, RchyOptimyx-visualized cell type hierarchy} on a bone marrow sample from a patient with AML. Cell population identification strategy used for SSC and CD45, with the CD34-enriched subset highlighted (e).
RchyOptimyx analysis showing CD34 enrichment (f). }
\label{fig:01}
\end{figure*}

We also computed the limits for $k$ on a hypothetical machine with 12GB of RAM for samples representative of mass cytometry (Fig. \ref{fig:01}c) and polychromatic flow cytometry (Fig. \ref{fig:01}d), both of which would be intractable for flowType-BF.
flowType and RchyOptimyx are now able, within the memory of a common workstation (12GB), to analyze 34-marker data. 

Finally, to demonstrate the importance of several partitions per marker, we applied flowType and RchyOptimyx to an acute myeloid leukemia sample from Flow Repository (ID FR-FCM-ZZYA) (Fig. \ref{fig:01}e-f).
CD34 is a stem-cell marker typically expressed on AML blast cells. These blasts are also known to have dimly positive CD45 expression and low SSC \cite{Vial2001}. 
By partitioning CD45 and SSC into four and three partitions, and naively running flowType and RchyOptimyx to search for CD34-enriched cell types, we were able to find that the SSC$^{low}$CD45$^{dim}$ cell type had a high proportion of CD34$^+$ cells, as expected.
This would not have been possible with only two partitions for each of CD45 and SSC.


\section{Lymphoma Diagnosis Quality Checking}
\subsection{Introduction}
We have designed and developed two methods flowType and RchyOptimyx in previous articles~\cite{Aghaeepour2012, Aghaeepour2012a, o2014enhanced}. The flowType method extracts cell populations as features from flow cytometry data, and then, a method such as ROC-AUC scores those features. Finally RchyOptimyx summarizes and visualizes important cell populations of the data.

Here our goal is to extend the method and create a framework for diagnosis quality checking purposes, \emph{i.e.} to report samples that might be misdiagnosed or require further investigation. For this, we first categorize the given samples into two groups according to their diagnosed label; then for each sample, we train our model using all samples except the selected one, and check how much our model agrees with the given diagnosis label, \emph{i.e.} leave one out cross validation scheme. At the end, we report the difference between the given labels and our predicted labels, and samples with the most difference between their predicted label and given label are candidates for further investigation by pathologists.
\subsection{Materials and Methods}
\subsubsection{Flow Cytometry Data}
Our data consists of two cohorts consisting Diffuse large B-cell lymphoma (DLBCL) and Follicular Lymphoma (FL) patients. Table~\ref{tbl:fcs-qa-cohorts} shows a summary of the two cohorts.

\begin{table}[!ht]
  \centering
  \begin{tabular}{crrr}
    Cohort & FL & DLBCL & Sum \\ \hline
    A & 49 & 22 & 71 \\
    B & ?? & ?? & ??
  \end{tabular}
  \caption{Cohorts A and B in numbers}
  \label{tbl:fcs-qa-cohorts}
\end{table}

Each cohort consists of three 8-color flow cytometry tubes. We include forward and side scatter values to the analysis, and treat them as biomarkers. Therefore each tube gives us a $10$ dimensional data for each patient. Table~\ref{tbl:fcs-qa-markers} presents markers on each tube for both cohorts A and B.

\begin{table}[!ht]
  \centering
  \begin{tabular}{rlm{7cm}}
    Cohort & Tube & Markers \\ \hline
    & Tube 1 & FS, SS, polyKappa, polyLambda, CD5, CD10, CD11c, CD20, CD3, CD19 \\
    A & Tube 2 & FS, SS, FMC-7, CD103, CD5, CD38, CD23, CD25, CD19, CD3 \\
    & Tube 3 & FS, SS, CD57, CD7, CD5, CD2, CD56, CD8, CD3, CD4 \\ \hline
     & Tube 1 &  \\
    B & Tube 2 &  \\
     & Tube 3 &  \\
  \end{tabular}
  \caption{Combination of markers in three tubes on each cohort. FS: Forward Scatter, SS: Side Scatter}
  \label{tbl:fcs-qa-markers}
\end{table}


\subsubsection{Cell Population Identification and Preprocessing}
We assume each marker divides cells into two populations, positive and negative. For each marker, a k-means clustering method automatically clusters cells into two groups. The group with a higher value of the corresponding marker is our positive, and the other group is our negative populations with regard to the marker. A threshold right in the middle of the two groups determines the two corresponding populations.

Then flowType counts the number of cells for each possible population considering determined thresholds. We then normalize these cell counts dividing them by total cell count to derive percentage of cells in each area. Having $M$ markers, we detect $3^M$ cell populations~\cite{Aghaeepour2012a}, and put them in a row, one row for each sample. At the end we normalize each column of the resulting matrix by estimating each column's mean ($\mu_i$) and standard deviation ($\sigma_i$), and then transform its values ($x$) according to Formula~\ref{frm:fcs-normal-transform}:

\begin{align}
  x^t := \frac{x - \mu_i}{\sigma_i}
  \label{frm:fcs-normal-transform}
\end{align}

For the purpose of diagnosis, we are interested mostly in cell populations that on average have more than only a few cells per sample. Therefore we discard cell populations which have a median value less than $0.05\%$ of total cell count. Such a preprocessing of the data results in a matrix for each tube. Each sample corresponds to one row in each of these matrices.

\subsubsection{Sample Classification}
We compare classification performance of several methods, which include support vector machines in three variants, gradient boosting classifier~\cite{friedman2002stochastic}, and a method which performed well in a previous benchmark competition, hereafter referred to as \emph{team21}\footnote{\url{http://www.ehu.eus/biologiacomputacional/team21_vilar}}~\cite{Aghaeepour2013Critical}. Table~\ref{tbl:fcs-qa-performances} presents a performance comparison between the abovementioned methods for our two cohorts.

\begin{table}[!ht]
  \centering
  \begin{tabular}{rcc}
     & \multicolumn{2}{c}{ROC AUC} \\ \cline{2-3}
    Method & Cohort A & Cohort B \\ \hline
    team21 & 0.5 & ? \\
    Gradient Boosting Classifier & 0.84 & ? \\
    SVM & 0.76 & ? \\
    $l1$-SVM-linear & 0.84 & ? \\
    $l2$-SVM-linear & 0.83 & ?
  \end{tabular}
  \caption{Method Performances - SVM: Support Vector Machine, $\{l1, l2\}$-SVM-linear: SVMs with a linear kernel which are penalized using an $l1$ or an $l2$ term respectively.}
  \label{tbl:fcs-qa-performances}
\end{table}

As shown in Table~\ref{tbl:fcs-qa-performances}, Gradient Boosting Classifier, $l1$-SVM-linear, and $l2$-SVM-linear all have comparable performances. Among these three methods, we choose $l2$-SVM-linear for the next steps for two reasons: proper handling of imbalanced data, and interpretability of its coefficient vector as features' importance. We choose $l2$-SVM-linear over $l1$-SVM-linear because $l1$ penalized SVM tends to be very sparse, meaning that it chooses only a few features as predictors. This results in usually choosing features that correspond to fewer cells and ignoring abondant cell populations.

Now we compute the difference between the output of the $l2$-SVM-linear method, and the desired output, \emph{i.e.} $-1$ and $1$. Note that the model's output is a real value, hence this results in non-integer values. Table~\ref{fig:fcs-qa-matchcount} show a few entries of the resulting table. The first three entries show an example of samples our method has mis-classified, and the last two entries are two samples for which our method agrees with diagnosis.

\begin{table}[!ht]
  \centering
  \begin{tabular}{llrr}
    Sample ID & Diagnosis Label & Target Value & Average Predicted Value \\ \hline
    F09-0939 & FL3 & 1 & -1.16 \\
    F09-1578 & FL12 & 1 & -1.05 \\
    F09-0628 & DLBCL & -1 & 0.80 \\
    F09-0578 & FL12 & 1 & 2.83 \\
    F09-1471 & DLBCL & -1 & -3.07
  \end{tabular}
  \caption{A sample of prediction values compared to target values. The target value is $-1$ for DLBCL samples and $1$ for FL samples. Average prediction value shows the average output of $l2$-SVM-linear method over three tubes.}
  \label{fig:fcs-qa-matchcount}
\end{table}


\subsubsection{Interpretation and Visualization}
A support vector machine with a linear kernel, uses a linear combination of input features to classify samples. Because we transformed and normalized input features before training our models, it is possible to directly interpret feature weights assigned by the model as importance of features.

Using a leave-one-out cross-validation scheme over $n$ samples, means we have $n$ different trained models for each tube, given a specific method. Therefore for every tube, we have $n$ different feature weights. We take the absolute value of the average feature weights over these $n$ models as our final feature score. RchyOptimyx then uses these feature scores to visualize features with higher scores. Figures~\ref{fig:fcs-qa-rchy1} and ~\ref{fig:fcs-qa-rchy2} show our analysis on cohorts A and B respectively.

\begin{figure}[!ht]
  \subfigure[Tube 1]{
    \includegraphics[width=.5\textwidth]{figs/fcs-qa/tube1}
    }
  \subfigure[Tube 2]{
    \includegraphics[width=.5\textwidth]{figs/fcs-qa/tube2}
    }
  \subfigure[Tube 3]{
    \includegraphics[width=.5\textwidth]{figs/fcs-qa/tube3}
    }
  \caption{RchyOptimyx analysis cohort A}
  \label{fig:fcs-qa-rchy1}
\end{figure}

\begin{figure}[!ht]
  %\includegraphics{}
  TODO
  \caption{RchyOptimyx analysis cohort B}
  \label{fig:fcs-qa-rchy2}
\end{figure}

Now we present a post-analysis of our methods, including an analysis of individual samples. Our goal is to give insight about why our models have mis-classified an individual sample. An oncologist/pathologist can use this part of the analysis as a hint for further investigation on the patient, or she can decide the mis-classification is due to overfitting or other errors in our models.

As already mentioned, we follow a leave one out cross validation scheme. This means we train a model for each sample $x_i$, using all samples but $x_i$. Let this model be referred to as $M_i$. Then we use coefficient vector of $M_i$ as an indicator of feature significance. Note that this is possible because we have re-scaled and normalized all the features. Now let the set $\mathcal{F}_{i,k}$ be the set of $k$ features with largest absolute values in feature vector of $M_i$. Then for each feature $f_j$ in $\mathcal{F}_{i,k}$:
\begin{itemize}
\item Let $\mathcal{P}_{DLBCL, f_j} := \{P_{f_j, s} | s \in DLBCL\}$
\item Let $\mathcal{P}_{FL, f_j} := \{P_{f_j, s} | s \in FL\}$
\end{itemize}

where $P_{f_j, s}$ is the cell count of the corresponding feature $f_j$ for sample $s$, and $DLBCL$ and $FL$ represent DLBCL diagnosed and FL diagnosed samples respectively. Then we draw the density plots corresponding to the two sets $\mathcal{P}_{DLBCL, f_j}$ and $\mathcal{P}_{DLBCL, f_j}$ over each other, and an indicator of where $f_i$ of $M_i$ stands over them. Figure~\ref{fig:fcs-qa-sample-density} shows an example result of this analysis.

\begin{figure}[!ht]
  \centering
  \includegraphics[width=.6\textwidth]{figs/fcs-qa/density-analysis/F09-0939_tube1_0202000120_density}
  \caption{Sample density analysis - TODO}
  \label{fig:fcs-qa-sample-density}
\end{figure}

A mis-classification occures when the given sample lays where the opposite class has higher density. For a given mis-classified sample $M_i$, and its corresponding set $F_{i,k}$, we draw the explained plots, and also their corresponding scatter plot of the cell population. Such a process results in plots shown in Figure~\ref{fig:fcs-qa-density-scatter-DLBCL} and~\ref{fig:fcs-qa-density-scatter-FL}.

\begin{figure}[!ht]
  \centering
  \caption{Actual Density-Scatter plots}
  \label{fig:fcs-qa-density-scatter-DLBCL}
\end{figure}

\begin{figure}[!ht]
  \centering
  \caption{Actual Density-Scatter plots}
  \label{fig:fcs-qa-density-scatter-FL}
\end{figure}

TODO: Check corresponding RchyOptimyx plots.



