
\begin{savequote}[.5\linewidth]
  ``Growth for the sake of growth is the ideology of the cancer cell.''
  \qauthor{- Edward Abbey}
\end{savequote}
\chapter{Introduction}
\label{ch:intro}
Cancer has been with human species throughout our 4000 years of history.
Although our understanding of cancer has changed drastically through time, its
treatment stays challenging and in many cases we have not been able to cure it
to this date. The immense frustration of dealing with cancer not only affects
the patients, but also the doctors, pathologists, and oncologists treating
those patients. Siddhartha Mukherjee in his book ``The Emperor of All
Maladies'' explains the feeling with these
words~\cite[prologue]{the-emperor-of-all-maladies}:

\begin{displayquote}
  ...
  
  There were seven such cancer fellows at this hospital. On paper, we seemed
  like a formidable force: graduates of five medical schools and four teaching
  hospitals, sixty-six years of medical and scientific training, and twelve
  postgraduate degrees among us. But none of those years or degrees could
  possibly have prepared us for this training program. Medical school,
  internship, and residency had been physically and emotionally grueling, but
  the first months of the fellowship flicked away those memories as if all of
  that had been child's play, the kindergarten of medical training.

  ...

  The stories of my patients consumed me, and the decisions that I made haunted
  me. \emph{Was it worthwhile continuing yet another round of chemotherapy on a
    sixty-six-year-old pharmacist with lung cancer who had failed all other
    drugs? Was it better to try a tested and potent combination of drugs on a
    twenty-six-year-old woman with Hodgkin's disease and risk losing her
    fertility, or to choose a more experimental combination that might spare
    it? Should a Spanish-speaking mother of three with colon cancer be enrolled
    in a new clinical trial when she can barely read the formal and inscrutable
    language of the consent forms?}
\end{displayquote}

I can better put this thesis in context by giving my personal perspective on
cancer research and diagnosis. I spent over a year researching in British
Columbia's Cancer Research Center (BCCRC) in Vancouver, Canada, which also
admitted patients for diagnosis and treatment. As a result, I worked closely
with oncologists who were diagnosing patients as well as cell biologists
researching fundamentals of cancer. Their experiences and the challenges they
were facing had a great impact on me and to a large extent shaped my research
for the next few following years. Then, once I moved to the Bioinformatics lab
in Max Planck Institute for Informatics in Saarbr\"ucken, Germany, I was better
equipped with required statistics and machine learning skills to tackle the
computational problems explained here. Following are a few examples of
challenges I saw people were facing in BCCRC:

\begin{itemize}
  \item Some patients enter the clinic carrying cancer type A, which is mild
    and does not require an aggressive treatment. Therefore they are put under
    the appropriate treatment while their condition is monitored through time.
    However, the disease in some of these patients develops into another type,
    let say type B, which is more aggressive and sometimes requires a harsher
    or a different treatment. Considering the fact that like many other
    diseases, cancer can be defeated best in its earliest stages, we could
    potentially achieve a better prognosis for these patients had we known
    their disease will develop into type B at a much earlier stage.

  \item Similar to the above issue, out of the many patients who go in
    remission, \emph{i.e} they seem free of cancer after the course of the
    treatment, some relapse with a cancer which is significantly more resistant
    to usual treatments compared to when they where originally diagnosed. This
    sometimes happens when a very small number of cells from the original
    cancer are or become resistant to the drugs and survive the treatment, but
    go undetected for a while in the tests and scans. It may take months for
    them to grow large enough to be detected again. Now the question is,
    looking back at the data of these patients, could we detect those cells, or
    something about the original cancer cells predicting the relapse, earlier
    during the treatment or even at the time of the original diagnosis?

  \item In BCCRC, people were also researching cancer by looking at the effects
    of different drugs and drug combinations targeting different genes. Some of
    my cell biologist friends, often taking recommendations from their
    supervisors, would choose a few genes and spend years investigating the
    role of those genes in the development of a particular cancer type. Of
    course they would do their best to choose the most relevant set of genes to
    their knowledge, but the task of choosing a few genes out of over 22k genes
    on the human genome is rather challenging and does not always lead to
    positive results and successful treatments. There is also a bias towards
    the genes which have been discovered earlier and have been studied more in
    depth. If the cancer happens to be related to one of the less studied
    genes, it usually can stay under the radar for while.
\end{itemize}

TODO: The two parts are not clear

These(TODO: ?) decisions are hard to make, and it does not help knowing that
our deep understanding of cancer is far from complete. The battle against
cancer has many fronts, including prevention, diagnosis, and treatment, all of
which benefiting from the advancements in understanding the disease. As a part
of the process, cancer researchers try to understand the disease in the lab,
and once their findings are confirmed, accepted by the community, and pass the
legal requirements, they are used by pathologists and oncologists in the
clinics. However, the diagnosis itself is also complex, challenging, and in
many cases not a definitive one. This is why sometimes doctors do not agree on
the exact diagnosis, and a counsel of experts is required for a better and more
reliable diagnosis and a treatment which hopefully results in a better
prognosis.

Cancer is a collection of extremely smart and complicated diseases. Although
they share many common characteristics, the same treatment does not result in a
similar prognosis. For example, two patients may come with two very similar
malignant tumors in their breasts. However, one of the patient's tumor grows in
response to estrogen, while the other one shows no reaction to estrogen. In
this case, a treatment which blocks estrogen receptors is very effective for
the first patient (ER+), while being completely ineffective against the second
patient who has an ER- subtype of breast cancer.

At the core of it, it comes to the fact that in normal cells there are
processes and checks put in place which define when and if the cell should
divide or die at a certain time or under certain conditions. Some of those
mechanisms act like an automated self destruct switch which is triggered if
something goes wrong in the cell. However, our cells are under constant stress
from the external factors which damage them, UV being one example, and
sometimes the damage to the cell affects those mentioned mechanisms and
disables them. This may lead the cell to divide uncontrollably, and become
cancerous. Another difference between normal cells and cancerous ones, is that
normal cells are capable of repairing most of the mutations happening on their
DNA, whereas those processes themselves are damaged in a cancerous cell. As a
result, the rate of mutation in cancerous cells is higher of a few orders of
magnitude compared to a normal healthy cell. The high rate of cell division in
combination with hyper mutation, makes cancerous cells very adaptive to their
environment, as well as against the drugs attacking them.

Cancer treatments are methods and substances which ideally target only the
cancerous cells and kill them. In a sense, they are poison, but ideally only to
cancer. However, cancer cells are derived from our own cells, and therefore it
is not easy to distinguish them from normal cells. The more difference between
the cancer cells and our normal healthy cells, the easier to target them; but
unfortunately not all cancer subtypes are easily distinguishable from other
cells for the purpose of treatment.

TODO: we don't say why we help and how

This thesis is an effort to design and develop computational methods to tackle
the abovementioned challenges. There are two main parts to this thesis. In the
first part, presented in Chapter~\ref{sec:fcs}, the focus is mostly on the
methods applicable directly in clinics, using the kind of data readily
available at the time of the diagnosis. The technology required to produce the
required data is not new and it has been in widespread use for decades all
around the world. The technology is called flow cytometry and the machine
producing the data is a flow cytometer. The machine takes certain measurements
from single cells, one by one, allowing us to study the presence and absence of
certain cell types in a single biopsy. It also allows us to sort and extract
certain cell types from the biopsy for further analysis on only those selected
cells. Although the technology has been around for a while, up until recently
researchers have been analyzing its data manually. This work presents methods
to automate some of those laborious and tedious tasks, while deriving new
information and insights about the data at the same time. As a byproduct, these
methods can be used to design tests which can be run on cheaper machines
available in poorer countries using the data from the more advanced machines,
and yet deliver a relatively similar performance.

TODO: chapter 3 doesn't seem new here.

TODO: be clear on what is done!

TODO: same paragraph as above

More specifically, one way pathologists study a given biopsy, is to find out
what kinds of cells are present in a given sample, and what types of
irregularities they posses. Each of these groups of cells is called a cell
population, and what is presented in Chapter~\ref{sec:fcs} is an effort to find
different ways of discovering meaningful cell populations in each given sample.
At the same time, we detect certain cell populations which appear to be
correlated with a specific disease, or are helpful differentiating between
two subtypes of a disease.

The other part of this thesis is closer to the bleeding edge of cancer research
compared to what happens in clinics. It uses some types of data, such as gene
expression profiles and DNA methylation data, which are more expensive to
acquire and are usually measured for the purpose of research. A gene expression
profile measures the activity of all 22k+ genes in a given sample, and a usual
DNA methylation profile measures the methylation level of about 450k sites on
the human DNA. These data, as well as others such as DNA, RNA, and protein
sequence data have been increasingly used by biologists and computational
biologists to better understand cell biology in many fields, including cancer
research. These data and methods have been so essential to our understanding of
cancer that the classification of some cancer types now depend on them. In some
cases, they(?, TODO) have shown us that two different classes of cancer are
indeed the same disease, only in different stages. Lymphoma and leukemia are
two good examples which used to be considered two different cancers, and now
put together in the \emph{lymphoid neoplasms} group.

TODO: bad place, move up?

TODO: what's the other one?

One goal of this thesis is to design computational methods capable of reporting
a list of promising genes that are influential in determining the cancer
subtypes, so that the same set of genes can be a better starting point for cell
biologists to choose and investigate. Doing so, we can also deliver a list of
genes specific to each patient, enabling the selection of a rather personalized
treatment for the patient.

Although there has been magnificent advancements in the field from the
computational perspective, these(?, TODO) are very hard problems. The curse of
dimensionality on top of the low number of samples compared to the number of
dimension in the data all result in a hard computational problem as further
explained in Section~\ref{sec:background}). Also, the fact that the data is
often affected by noise and batch effects doesn't help the case neither, which
we cover in more detail in Section~\ref{sec:adaptive-learning}. Another
challenging factor is that cancer, and even a single cancer tumor, is
heterogeneous and different cases show very different varying genetic profiles.
Historically this has lead to further classification of cancer, and at times,
changing the classification and merging some classes all together as mentioned
above.

All of the above facts and arguments motivated me to focus on investigating
interpretable machine learning models so that:
\begin{itemize}
\item The result of the investigation and application of the model on the data
  results in better understanding of the disease, by giving cell biologists
  hints about where they can put their focus.
\item For any given data from a single patient, the model can give some
  information about the potential underlying cause of the disease which in
  combination with the fact that we often times have several treatment options
  for the same disease can be used in the diagnosis phase to improve the
  treatment and prognosis. (TODO: one sentence!)
\item For a trained model, give an overall explanation on how the model is
  working so that it can better be trusted and understood by the people in
  clinics.
\end{itemize}

In the following chapters, some of the basics required to follow the later
sections, covering some background in the related parts of machine learning and
biology are covered in chapter~\ref{sec:background}. In chapter~\ref{sec:fcs}
we focus on flow cytometry data and explain the design and implementation of a
few pieces which together make an end to end pipeline to analyze such data, and
apply that to some specific lymphoma subtypes. Then we continue in
chapter~\ref{sec:adaptive-learning} with the analysis of mostly DNA methylation
data and design some adaptable and interpretable models with an eye on
personalized medicine.

TODO: what are the concrete research questions?
